### DISCUSSION TITLE
7 Axis Machine

### DISCUSSION THREAD

**Discussion by cr59cummins** [grblHAL/core#428] [Q&A] [answered]
I am in the process of building a machine that is going to have 7 axis. It will be a two station welder and the machine will not be programmed using any cad. It will be jogged to a location, position recorded and then jogged to the next position and recorded. It is not likely that more than 3 axis will move at the same time. The programs will be for our products and probably won't change for years after being setup. We will run one program at a time and it will run repeatedly.
I know there are plenty of ways to do this and even more robust industrial systems but I have experience with GRBL and already have an iPad app and server that tracks parts processed and can be monitored / controlled from the office. I am trying to use as much of my already running system as possible so I don't need a massive overhaul.
I know I can buy industrial cnc controls that can handle 8+ synchronous axis but the goal is to have a headless design where the operator can power it on and go.

The axis letters I am using in the following description are just letters to help readers visualize the machine. The letters of axis do not matter to me in use because I can change the iPad app to display the letters the way I like to see them.

There will be two parallel axis, one for each to move the parts in and out of the welding area independantly (X1, X2). Each of these will have a rotary axis (A, B). There is an axis (Y) perpendicular to the parallel ones that will carry a (Z) and on the Z axis there will be another 0 to ~50 degree (C).

From looking at the Driver capabilities page, it looks like the max axis is 6 unless I missed something. I am posting to see if anyone knows of a way to have 7 controlled axis. If not, I am considering using the the spindle PWM out to tell another controller where to place the 0 to 50 degree axis. (I don't need spindle speed control)

Another option is to have two slave GRBL with one master that controls the flow of GCode to each slave. Slave 1 (X1, X2) would be positional and slave 2 would have the rest. This would be slower runtime since the axis would need to finish moving before the master could tell the second slave to start. It would also be a lot more to program.

Sorry for the long explanation. I was trying to get enough information out for you to understand what I am trying to accomplish but there is no way for me to get you everything that's part of this project.

I am also considering not using GRBL but I like the synchronous and the GCode flow.

**Comment by cr59cummins** [comment]
I should have also specified that I do not need a predesigned board. I can create the board around the processor.

**Comment by terjeio** [comment, marked as answer]
The core supports up to 8 axes, most drivers up to 6 for now.
The exception is the [STM32F7xx](https://github.com/grblHAL/STM32F7xx) and [STM32H7xx](https://github.com/dresco/STM32H7xx) drivers which supports 8. The Nucleo 144 boards has enough I/O for this and the [reference maps](https://github.com/grblHAL/STM32F7xx/blob/main/Inc/reference_map.h) enables all. The recently added [BTT Octopus MAX 3](https://bigtreetech.github.io/docs/Octopus.html) board seemingly can handle 8 axes - but the map file is for up to 6, the missing [pin definitions](https://github.com/dresco/STM32H7xx/blob/master/Inc/btt_octopus_max_map.h) is easy to add.

**Reply by dresco** [reply]
Just noticed the above link for the Octopus board wasn't for the MAX, correct one [here](https://bigtreetech.github.io/docs/Octopus%20MAX%20EZ.html)..

**Reply by dresco** [reply]
Something else to note & a question for Terje. The MAX board only has 6 limit switch inputs, is it possible to add additional axis without limit pins? (It actually has connections for 10 Trinamic drivers).

**Reply by cr59cummins** [reply]
Thank you very much for the replies, I had not seen those boards before. I was planning to use some Teknic Clearpath SDSK servos that I had left over from a previous build. It looks like the Max cannot output step/dir only spi or uart unless there is a plugin driver that can convert it that I did not see. I might just use steppers since this board looks very handy. Would the Octopus Pro work since it can output step/dir?

**Reply by terjeio** [reply]
> Would the Octopus Pro work since it can output step/dir?

It can, but you will have to get the signals out of the new style driver sockets if you are going to use external drivers. SPI or UART is used to configure Trinamic drivers, not to run the motors.
If using external drivers a new board map for the [Octopus](https://bigtreetech.github.io/docs/Octopus.html) might be simpler to get working since it has Polulu style stepper driver sockets. Earlier I made a [buffer for external drivers](https://github.com/terjeio/CNC_Boosterpack#bonus-design-polulu-style-motor-driver-dummy) that plugs into these.

**Comment by terjeio** [comment]
>  is it possible to add additional axis without limit pins?

Yes. Homing has to be done manually if needed, and the motors cannot be used for auto squared axes. Or a plugin [like this](https://github.com/grblHAL/ESP32/blob/master/main/boards/BlackBoxX32.c) can be added that reroutes the limit signal from another axis when homing. Note that the two axes cannot then be homed simultaneously.
